source/simulator/core.cc

```diff
@@ -2147,6 +2167,286 @@ namespace aspect
+               */
+              do
+                {
+                  current_linearization_point = backup_linearization_point;
+
+                  LinearAlgebra::BlockVector search_direction = solution;
+
+                  search_direction *= lambda;
+
+
+                  current_linearization_point.block(introspection.block_indices.pressure) += search_direction.block(introspection.block_indices.pressure);
+                  current_linearization_point.block(introspection.block_indices.velocities) += search_direction.block(introspection.block_indices.velocities);
+
+                  assemble_newton_stokes_matrix = rebuild_stokes_preconditioner = false;
+                  rebuild_stokes_matrix = true;
+
+                  assemble_stokes_system();
+
+                  test_velocity_residual = system_rhs.block(introspection.block_indices.velocities).l2_norm();
+                  test_pressure_residual = system_rhs.block(introspection.block_indices.pressure).l2_norm();
+                  test_residual = std::sqrt(test_velocity_residual * test_velocity_residual
+                                            + test_pressure_residual * test_pressure_residual);
+
+                  if (test_residual < (1.0 - alpha * lambda) * residual
+                      ||
+                      line_search_iteration >= parameters.max_newton_line_search_iterations
+                      ||
+                      use_picard)
+                    {
+                      pcout << "      Total relative residual after nonlinear iteration " << nonlinear_iteration+1
+                            << ": " << residual/initial_residual << ", norm of the rhs: " << test_residual
+                            << ", newton_derivative_scaling_factor: " << newton_handler->get_newton_derivative_scaling_factor() << std::endl;
+                      break;
+                    }
+                  else
+                    {
+
+                      pcout << "   Line search iteration " << line_search_iteration << ", with norm of the rhs "
+                            << test_residual << " and going to " << (1.0 - alpha * lambda) * residual
+                            << ", relative residual: " << residual/initial_residual << std::endl;
+
+                      /**
+                       * The line search step was not sufficient to decrease the residual
+                       * enough, so we take a smaller step to see if it improves the residual.
+                       */
+                      lambda *= (2.0/3.0);// TODO: make a parameter out of this.
+                    }
+
+                  line_search_iteration++;
+                  Assert(line_search_iteration <= parameters.max_newton_line_search_iterations,
+                         ExcMessage ("This tests the while condition. This condition should "
+                                     "actually never be false, because the break statement "
+                                     "above should have caught it."));
```

Comment:
Well, I don't think this assert can every be triggered unless someone changed the code which is about 25 lines above this line (the code where it basically states if line_search_iteration >= parameters.max_newton_line_search_iterations -> break (note that there is a line_search_iteration++; in between). So what I basically say in this assert is that you changed the code just above this assert in such a way that it is broken. So it was meant more for developers than for users. 

I personally think that we should remove this assert, because I doubt  how useful this assert is in reality for developers, because they might be expected to know what they are doing before changing core.cc and test what they changed anyway. Or do you want these kind of securities in the code?